[FLINK-4035] Add a streaming connector for Apache Kafka 0.10.x

rmetzger opened the pull request:
This pull request subsumes https://github.com/apache/flink/pull/2231.

Compared to #2231, I've based the connector on the existing 0.9 code (by extending it, reducing the amount of copied code), added a test case for the timestamp functionality and updated the documentation.

StephanEwen commented:
Just looked over this briefly.
What struck me first is that this again uses the dirty trick of adding a dependency to "Flink Kafka 0.9" and then transitively excluding "Kafka 0.9". Is there a nicer way to solve this?

tzulitai commented on flink-streaming-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcher.java:
```diff
@@ -207,15 +207,14 @@ public void restoreOffsets(HashMap<KafkaTopicPartition, Long> snapshotState) {
+	 * @param kafkaRecord The original Kafka record
 	 */
-	protected final void emitRecord(T record, KafkaTopicPartitionState<KPH> partitionState, long offset) {
+	protected <R> void emitRecord(T record, KafkaTopicPartitionState<KPH> partitionState, long offset, R kafkaRecord) throws Exception {
```
Is there a reason we need to have an extra `kafkaRecord`? It doesn't seem to be used in the function.

tzulitai commented on flink-streaming-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka010Fetcher.java:
```diff
@@ -0,0 +1,91 @@
+	 * Emit record Kafka-timestamp aware.
+	 */
+	@Override
+	protected <R> void emitRecord(T record, KafkaTopicPartitionState<TopicPartition> partitionState, long offset, R kafkaRecord) throws Exception {
```
Just realized here the reason for the new `kafkaRecord`.

However, would it be better to add an `recordTimestamp` argument instead of passing the original Kafka record?
If we don't have such a timestamp (for 0.8, 0.9), a `null` can be accepted. In `AbstractFetcher#emitRecord()`, we check if the value is `null` or not and correspondingly call `collect()` or `collectWithTimestamp`. The same goes for passing the `Long.MIN_VALUE` / timestamp to `emitRecordWithTimestampAndPeriodicWatermark` and `emitRecordWithTimestampAndPunctuatedWatermark `.

IMHO, I think this way the new code will be more meaningful and less confusing for the base `AbstractFetcher`, and also we won't need to override `emitRecord` in 0.10.

rmetzger replied:
I agree that this is confusing at the `AbstractFetcher` level. I'll look into that as well.

tzulitai commented on flink-streaming-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer010.java:
```diff
@@ -0,0 +1,198 @@
+ * Implementation note: This Producer wraps a Flink Kafka 0.9 Producer, overriding only
+ * the "processElement" / "invoke" method.
+ */
+public class FlinkKafkaProducer010<T> extends StreamSink<T> {
```
Overall, I think the solution here seems a bit too "hacky" to me. Here's a few disadvantages I see: 1) the code in `processElement()` of this class is still quite a bit duplicate of the `invoke()` in `FlinkKafaProducerBase`. 2) with `writeToKafka()` we need to have a different usage style for the Kafka producer in 0.10, compared to previous versions.

If I'm correct, I guess its due to the fact that we don't have access to the embedded record timestamp in the provided `next` in the usual `invoke`?

I'm wondering whether or not a cleaner solution is to introduce a `serializeTimestamp(T element)` method in the `KeyedSerializationSchema`, and we can simply use that to make `FlinkKafkaProducerBase#invoke()` more general by replacing instantiation of `ProducerRecord`s with an abstract version-specific method. Then, `FlinkKafkaProducer010` can simply extend `FlinkKafkaProducer09`.

We'll need a migration plan though if we're going to change the serialization schema interface. If we're going for it, might as well migrate the `KeyedDeserializationSchema` to include the timestamp too (give `null` for 0.8, 0.9).
Otherwise, I think the solution here is ok for a short term solution. What do you think?

tzulitai replied:
But on the other hand, the downside for the approach above would be that users always have to explicitly extract the timestamp again (even if the record already has an embedded timestamp) to emit them to Kafka.

rmetzger replied:
Exactly. The problem is that there is no way of accessing the system's event time record.

Maybe I can try and build something like a hybrid producer that works with both invocation methods.

tzulitai commented on flink-streaming-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer010.java:
```diff
@@ -0,0 +1,198 @@
+	/**
+	 * Configuration object returned by the writeToKafka() call.
+	 */
+	public static class FlinkKafkaProducer010Configuration<T> extends DataStreamSink<T> {
```
I wonder if we can move the config setter methods into `FlinkKafkaProducer010`, and let `writeToKafka` return the created 0.10 producer instead of instantiating a new `FlinkKafkaProducer010Configuration`? My reasoning is that I find it a bit strange to set these config values on a separate config object, but not "applying" it anywhere. I think users would normally expect they have to "apply" a completed config object somewhere to take effect. So, I have something like this in mind:

In constructor of `FlinkKafkaProducer010`
```
private FlinkKafaProducer010(...) {
  super(new FlinkKafkaProducer09<>(...));
  this.wrapped09Producer = (FlinkKafkaProducer09<T>) userFunction; // wrapped09Producer as FlinkKafkafProducer010's class field
}
```
Then FlinkKafkaProducer010 can directly have the setter methods:
```
public void setFlushOnCheckpoint(boolean flush) {
    this.wrapped09producer.setFlushOnCheckpoint(flush);
}

public void setLogFailuresOnly(boolean logFailuresOnly) {
	this.wrapped09producer.setLogFailuresOnly(logFailuresOnly);
}

public void setWriteTimestampToKafka(boolean writeTimestampToKafka) {
	this.writeTimestampToKafka = writeTimestampToKafka;
}
```

Then users can use the producer like this:
```
FlinkKafkaProducer010 producer = FlinkKafkaProducer010.writeToKafka(...)
producer.setLogFailuresOnly(...)
producer.setFlushOnCheckpoint(...)
producer.setWriteTimestampToKafka(...)
```

Although still a bit different from the usage patterns of previous versions, it looks a bit more similar compared to what the PR has now. I'm not sure if I may be missing something that lead to choosing to have a separate `FlinkKafkaProducer010Configuration`. Is there any particular reason?

rmetzger replied:
I'll update it.

tzulitai commented on docs/apis/streaming/connectors/kafka.md:
```diff
@@ -291,3 +301,35 @@ higher value.
+{% highlight java %}
+FlinkKafkaProducer010.FlinkKafkaProducer010Configuration config = FlinkKafkaProducer010.writeToKafka(streamWithTimestamps, topic, new SimpleStringSchema(), standardProps);
+config.setWriteTimestampToKafka(true);
+{% endhighlight %}
```
I find the usage pattern of this a bit unfamiliar. I've explained this in inline comments of the `FlinkKafkaProducer010` class.

tzulitai commented:
Left a few comments on some high-level design choices for a first review. Mostly on `FlinkKafkaProducer010`, I wonder if there are other better possibilities over there?

rmetzger commented:
@StephanEwen: The explicit exclude is actually not needed, because the kafka version defined in the connector has precedence over transitive kafka versions.
```
[INFO] --- maven-dependency-plugin:2.8:tree (default-cli) @ flink-connector-kafka-0.10_2.10 ---
[INFO] org.apache.flink:flink-connector-kafka-0.10_2.10:jar:1.2-SNAPSHOT
[INFO] +- org.apache.flink:flink-connector-kafka-0.9_2.10:jar:1.2-SNAPSHOT:compile
[INFO] |  \- org.apache.flink:flink-connector-kafka-base_2.10:jar:1.2-SNAPSHOT:compile
[INFO] +- org.apache.kafka:kafka-clients:jar:0.10.0.1:compile
[INFO] |  +- net.jpountz.lz4:lz4:jar:1.3.0:compile
[INFO] |  \- org.xerial.snappy:snappy-java:jar:1.1.2.6:compile
```

rmetzger commented:
Thank you for the review @tzulitai. I'll try to find some time soon to look into your comments in detail.

eliaslevy commented:
This may be the wrong place to bring this up, but as you are discussing changes to the Kafka connector API, I think it is worth bring it up.  

As I've pointed out elsewhere, the current connector API makes it difficult to make use of Kafka native serializer or deserializer (`org.apache.kafka.common.[Serializer, Deserializer]`), which can be configured via the Kafka client and producer configs.  

The connector code assumes that `ConsummerRecord`s and `ProducerRecord`s are both parametrized as `<byte[], byte[]>`, with the Flink serdes performing the conversion to/from `byte[]`.  This makes it difficult to make use of Confluent's `KafkaAvroSerializer` and `KafkaAvroDecoder`, which make use of their [schema registry](http://docs.confluent.io/3.0.0/schema-registry/docs/serializer-formatter.html#serializer).

If you are going to change the connector API, it would be good to tackle this issue at the same time to avoid future changes.  The connector should allow the type parametrization of the Kafka consumer and producer, and should make use of a pass through Flink serde by default.

rmetzger commented:
@eliaslevy, I assume you are referring to https://issues.apache.org/jira/browse/FLINK-4050. 
Its good that you are mentioning the issue again, so I can move it a bit up on my TODO list.

I would personally prefer to first add the Kafka 0.10 module in this pull request and then resolve FLINK-4050 independently. I know that this might lead to a little bit of duplicate work on the Kafka 0.10 code, but on the other hand its easier to discuss one issue at a time :)

eliaslevy commented:
@rmetzger that's the one. NP.  I realize breaking it up makes things easier.  I just thought I'd mention it.

rmetzger commented:
@tzulitai I've addressed your comments.
The Producer is now "hybrid": you can use it with both invocation methods.
The AbstractFetcher now accepts a long timestamp instead of a record.

tzulitai commented on flink-streaming-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer010.java:
```diff
@@ -0,0 +1,399 @@
+ *  Pre Kafka 0.10 producers only follow approach (a), allowing users to use the producer using the
+ *  DataStream.addSink() method.
+ *  Since the APIs exposed in that variant do not allow accessing the the timestamp attached to the record
+ *  the Kafka 0.10 producer has a section invocation option, approach (b).
```
'section' --> second?

tzulitai commented on flink-streaming-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer010.java:
```diff
@@ -0,0 +1,399 @@
+	 * @param serializationSchema User defined (keyless) serialization schema.
+	 * @param producerConfig Properties with the producer configuration.
+	 */
+	public static <T> FlinkKafkaProducer010Configuration writeToKafka(DataStream<T> inStream,
```
Now with hybrid invocation methods, would it be reasonable to name this as `writeToKafkaWithTimestamps` so that it's more meaningful and differentiable from the usual invocation?

rmetzger replied:
Its a good idea, I'll rename the methods

tzulitai commented on flink-streaming-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer010.java:
```diff
@@ -0,0 +1,399 @@
+	 */
+	public FlinkKafkaProducer010(String topicId, SerializationSchema<T> serializationSchema, Properties producerConfig, KafkaPartitioner<T> customPartitioner) {
+		this(topicId, new KeyedSerializationSchemaWrapper<>(serializationSchema), producerConfig, customPartitioner);
+
```
nit: unnecessary empty line

tzulitai commented on flink-streaming-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka010Fetcher.java:
```diff
@@ -60,12 +60,17 @@ protected void assignPartitionsToConsumer(KafkaConsumer<byte[], byte[]> consumer
-	protected <R> void emitRecord(T record, KafkaTopicPartitionState<TopicPartition> partitionState, long offset, R kafkaRecord) throws Exception {
-		long timestamp = ((ConsumerRecord) kafkaRecord).timestamp();
+	protected void emitRecord(T record, KafkaTopicPartitionState<TopicPartition> partitionState, long offset, long timestamp) throws Exception {
 		if (timestampWatermarkMode == NO_TIMESTAMPS_WATERMARKS) {
```
Is it possible to let `AbstractFetcher#emitRecord` determine whether to call `collectWithTimestamp` or `collect` depending on the provided timestamp (== `Long.MIN_VALUE`)? Then, we won't need to override the base `emitRecord` here, correct?

rmetzger replied:
Yes, I'll do that.

tzulitai commented on flink-streaming-connectors/flink-connector-kafka-0.8/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/SimpleConsumerThread.java:
```diff
@@ -376,7 +376,7 @@ else if (partitionsRemoved) {
 							}
 							
-							owner.emitRecord(value, currentPartition, offset, msg);
+							owner.emitRecord(value, currentPartition, offset, Long.MAX_VALUE);
```
I think this is supposed to give a `Long.MIN_VALUE` instead of MAX?

tzulitai commented on flink-streaming-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka09Fetcher.java:
```diff
@@ -262,6 +262,10 @@ public void run() {
 
+	// Kafka09Fetcher ignores the timestamp.
+	protected void emitRecord(T record, KafkaTopicPartitionState<TopicPartition> partition, long offset, ConsumerRecord consumerRecord) throws Exception {
+		emitRecord(record, partition, offset, Long.MAX_VALUE);
```
Same here: I think this is supposed to give a `Long.MIN_VALUE` instead of MAX?

tzulitai commented on flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcherTimestampsTest.java:
```diff
@@ -135,45 +135,45 @@ public void testPeriodicWatermarks() throws Exception {
-		fetcher.emitRecord(15L, part2, 3L, new ConsumerRecord<byte[], byte[]>(testTopic, 13, new byte[]{0}, 3L));
+		fetcher.emitRecord(13L, part2, 2L, Long.MAX_VALUE);
+		fetcher.emitRecord(14L, part2, 3L, Long.MAX_VALUE);
+		fetcher.emitRecord(15L, part2, 3L, Long.MAX_VALUE);
```
Same here: I think these are supposed to give a `Long.MIN_VALUE` instead of MAX?

tzulitai commented:
Thanks Robert for addressing my comments :)

Overall, I like the new hybrid producer approach. However, I'm still curious whether or not it is possible / reasonable to drop the `FlinkKafkaProducer010Configuration` return type of invocation (b), and let both invocation methods return `FlinkKafkaProducer010` instead. So,

```
FlinkKafkaProducer010 kafka = new FlinkKafkaProducer010(...)
// or FlinkKafkaProducer010 kafka = FlinkKafkaProducer010.writeToKafkaWithTimestamps(...) for timestamp support

// setter config methods directly done on the FlinkKafkaProducer010 instance regardless of (a) or (b)
kafka.setLogFailuresOnly(true)
kafka.setFlushOnCheckpoint(true)
kafka.setWriteTimestampToKafka(true) // would not have effect if original invocation method (a) was used
```

But we'll need to be bit hacky in `invokeInternal(element, elementTimestamp)`, something like only letting the given `timestamp` to `ProducerRecord` be non-null if `writeTimestampToKafka && elementTimestamp != Long.MIN_VALUE`.

What do you think?

nemccarthy commented:
+1 for this pr

cjstehno commented:
Any thoughts on when this might make it into a release? We are having issues running Flink with Kafka 0.10 and would like to have an idea of whether we can/should wait for this or pull it and try building our own. Thanks.

tzulitai commented:
Looks like we need to rebase this PR on the recently merged Kerberos support.

tzulitai commented:
@cjstehno I would expect this to be in the 1.2.0 major release, which would probably be ~2 months from now according to Flink's past release cycle. The Flink community usually doesn't release major new features like this between minor bugfix releases.

rmetzger commented:
I'm currently working on rebasing the PR and addressing the comments.

rmetzger commented:
@tzulitai I addressed all your comments except the one relating `FlinkKafkaProducer010Configuration`: I had a quick offline discussion with @StephanEwen about the issue and he suggested to add the timestamp to the regular sink interface.
But I would like to make that change separate from this one, and merge the Kafka 0.10. support as-is. This will make it easier for people to try it out now and provide us with feedback. Also, I think some other Kafka related pull requests are blocked on this one.

@tzulitai could you do a final pass over the changes. If you agree, I'd like to merge it afterwards.

tzulitai commented:
@rmetzger Thanks for addressing the comments! Did a final pass, and the changes look good to me.
I agree with merging the connector as is. Adding the timestamp to the regular sink interface seems like a good long term solution.

+1 to merge once travis turns green ;)

tzulitai commented on flink-streaming-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer010.java:
```diff
@@ -0,0 +1,398 @@
+																					   String topicId,
+																					   KeyedSerializationSchema<T> serializationSchema,
+																					   Properties producerConfig,
+																					   KafkaPartitioner<T> customPartitioner) {
```
Checkstyle failed for these few lines here, have leading spaces :P

tzulitai commented on flink-streaming-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer010.java:
```diff
@@ -83,11 +83,11 @@
-										KeyedSerializationSchema<T> serializationSchema,
-										Properties producerConfig) {
-		return writeToKafka(inStream, topicId, serializationSchema, producerConfig, new FixedPartitioner<T>());
+	public static <T> FlinkKafkaProducer010Configuration writeToKafkaWithTimestamps(DataStream<T> inStream,
```
Add the generic type parameter `T` to `FlinkKafkaProducer010Configuration` here too?

tzulitai commented on flink-streaming-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer010.java:
```diff
@@ -102,11 +102,11 @@
-										SerializationSchema<T> serializationSchema,
-										Properties producerConfig) {
-		return writeToKafka(inStream, topicId, new KeyedSerializationSchemaWrapper<>(serializationSchema), producerConfig, new FixedPartitioner<T>());
+	public static <T> FlinkKafkaProducer010Configuration writeToKafkaWithTimestamps(DataStream<T> inStream,
```
Add the generic type parameter `T` to `FlinkKafkaProducer010Configuration` here too?

tzulitai commented:
Just found some minor issues that can be fixed when merging.

rmetzger commented:
Thank you for the review. I'll address your comments, rebase again, test it, and if it turns green merge it ;)

rmetzger commented:
Tests are running https://travis-ci.org/rmetzger/flink/builds/166441047
